pkg/certgenerator/v1beta1/generator.go

```diff
@@ -0,0 +1,215 @@
+	errServiceNotFound      = errors.New("unable to locate controller service")
+	errCertCheckFail        = errors.New("failed to check if certs already exist")
+	errCreateCertFail       = errors.New("failed to create certs")
+	errCreateCertSecretFail = errors.New("failed to create secret embedded certs")
+	errInjectCertError      = errors.New("failed to inject certs into WebhookConfigurations")
+)
+
+// CertGenerator is the manager to generate certs.
+type CertGenerator struct {
+	namespace   string
+	serviceName string
+	secretName  string
+	kubeClient  client.Client
+	certsReady  chan struct{}
+
+	certs             *certificates
+	fullServiceDomain string
+}
+
+var _ manager.Runnable = &CertGenerator{}
+var _ manager.LeaderElectionRunnable = &CertGenerator{}
+
+func (c *CertGenerator) Start(ctx context.Context) error {
+	if err := c.generate(ctx); err != nil {
+		return err
+	}
+	// Close a certsReady means start to register controllers to the manager.
+	close(c.certsReady)
+	return nil
+}
+
+func (c *CertGenerator) NeedLeaderElection() bool {
+	return true
+}
+
+// AddToManager adds the cert-generator to the manager.
+func AddToManager(mgr manager.Manager, config configv1beta1.CertGeneratorConfig, certsReady chan struct{}) error {
+	return mgr.Add(&CertGenerator{
+		namespace:   consts.DefaultKatibNamespace,
+		serviceName: config.WebhookServiceName,
+		secretName:  config.WebhookSecretName,
+		kubeClient:  mgr.GetClient(),
+		certsReady:  certsReady,
+	})
+}
+
+// generate generates certificates for the admission webhooks.
+func (c *CertGenerator) generate(ctx context.Context) error {
+	controllerService := &corev1.Service{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.serviceName, Namespace: c.namespace}, controllerService); err != nil {
+		return fmt.Errorf("%w: %v", errServiceNotFound, err)
+	}
+
+	certExist, err := c.isCertExist(ctx)
+	if err != nil {
+		return fmt.Errorf("%w: %v", errCertCheckFail, err)
+	}
+	if !certExist {
+		c.fullServiceDomain = strings.Join([]string{c.serviceName, c.namespace, "svc"}, ".")
+
+		if err = c.createCert(); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertFail, err)
+		}
+		if err = c.updateCertSecret(ctx); err != nil {
+			return fmt.Errorf("%w: %v", errCreateCertSecretFail, err)
+		}
+	}
+	if err = c.injectCert(ctx); err != nil {
+		return fmt.Errorf("%w: %v", errInjectCertError, err)
+	}
+	return nil
+}
+
+// isCertExist checks if a secret embedded certs already exists.
+// For example, it will return true if the katib-controller is created with enabled leader-election
+// since another controller pod will create the secret.
+func (c *CertGenerator) isCertExist(ctx context.Context) (bool, error) {
+	secret := &corev1.Secret{}
+	if err := c.kubeClient.Get(ctx, client.ObjectKey{Name: c.secretName, Namespace: c.namespace}, secret); err != nil {
+		return false, err
+	}
+	key := secret.Data[serverKeyName]
+	cert := secret.Data[serverCertName]
+	if len(key) != 0 && len(cert) != 0 {
```

Comment:
Should we throw an error if some of the values are missing ?
Whether `tls.key` or `tls.crt`

Reply:
Instead of throwing an error, should we create new certs and then update Secret?
If `tls.key` or `tls.crt` is empty, I guess other controllers or users accidentally removed the value.
@andreyvelich I think it would be worth regenerating certs. WDYT?

Reply:
Makes sense, we need to change this check to `len(key) != 0 || len(cert) != 0`, right?

Reply:
Currently implementation, if either `key` or `cert` is empty, it will return false. This means the cert-generaotr generates new certs in the case either `key` or `cert` is empty. So I think we need not to change this check to `len(key) != 0 || len(cert) != 0`.

https://github.com/kubeflow/katib/blob/eac65bf8e31fe4d3a2c4dbb20be30d600c086157/pkg/certgenerator/v1beta1/generator.go#L124-L133

If I misunderstand, please correct me.

Reply:
Oh, sorry, I just missed it :)
Sounds good